### DISCUSSION TITLE
Converting strings to docs using an custom entity ruler model slow

### DISCUSSION THREAD

**Discussion by bhargav0795** [explosion/spaCy#8017] [Help: Coding & Implementations] [answered]
## How to reproduce the behavior
I built a Rule-based model using an entity ruler with 40k multi-words patterns as follows
{"label":"JT","pattern":[{"LOWER":"chief"},{"LOWER":"executives"}]}
{"label":"JT","pattern":[{"LOWER":"chief"},{"LOWER":"sustainability"},{"LOWER":"officers"}]}

What I'm trying to do?
The idea is to identify job titles from a string. I have a dataset of 140k strings. The process occurs in two phases, the first step involves identifying necessary data points from the string and break the string into multiple substrings, and then in the next step each substring is converted to a spacy doc object, and the job titles are extracted and stored.
`nlp.pipe()` function is invoked once per string. On average, a string is divided into 4 substrings. Average size of a substring is 850 words.

**Code**
`nlp.pipe(df.iloc[:10]['txt'], 
disable = ["tok2vec", "tagger", "parser", "attribute_ruler", "lemmatizer"],
n_process = len(df.iloc[:10]['txt'])-1, 
batch_size = 1)`

**Benchmarks**
To process 10 sub-strings 
```
%%timeit

4.86 s ± 93.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Is there a way to optimize the process? 

## Your Environment
* **spaCy version:** 2.3.0
* **Platform:** Linux-4.15.0-142-generic-x86_64-with-Ubuntu-18.04-bionic
* **Python version:** 3.6.9

**Comment by polm** [comment, marked as answer]
Sorry to hear you're having trouble with this.

> nlp.pipe() function is invoked once per string.

Ah, that could be a problem. That's not the way `nlp.pipe` is intended to be used, and could be slower than calling `nlp(text)`. I'm a little unclear if the argument to `nlp.pipe` is a single string or a list of four strings, but either way you should be batching up *all* your text and passing it to the pipe - if you're passing just one or a few strings it isn't going to help you. 

Since you have a lot of patterns and they look pretty simple I would also recommend you look at using the [PhraseMatcher][] instead of the Matcher, which can be faster. Be sure to look at how to specify an attribute (in your case `LOWER`) for it to use.

[PhraseMatcher]: https://v2.spacy.io/usage/rule-based-matching#phrasematcher

What model are you using?

Also, you're disabling the `attribute_ruler`, but I don't think spaCy 2 had attribute rulers.

**Reply by adrianeboyd** [reply]
A few notes:

* `nlp` vs. `nlp.pipe` doesn't make a difference for the entity ruler itself; however, if you're using a `tagger` and then `TAG`, `LEMMA`, or other attributes in your patterns, then `pipe` would speed things up
* if you're only using `ORTH` or `LOWER` (not `LEMMA`, `TAG`, or other attributes that require additional components), you can use `nlp = spacy.blank("en")` (or whichever language) to have a pipeline with just a tokenizer; then you don't need to disable anything
* what Paul means is to use phrase patterns with the entity ruler with `phrase_matcher_attr="LOWER"` in case all your patterns are token patterns with `LOWER` -- this will be much much faster because it uses the `PhraseMatcher` internally, but the entity ruler handles the internal details for the `PhraseMatcher` for you

In the example above, if you're using a pretrained pipeline downloaded with `spacy download`, then you've accidentally (?) left `ner` enabled, which will also slow down this processing. (I'm not sure whether your intended pipeline should be only `["entity_ruler"]` or [`"entity_ruler", "ner"]`, in which case this is totally fine and you should rewrite it to use `nlp.pipe` with all your texts or least larger batches of texts.)

**Comment by bhargav0795** [comment]
Firstly, thanks for the suggestions!
I've edited the entity ruler to now use phrase matcher and updated it as follows:

```
nlp_updated_implementation= spacy.blank("en")
patterns = [{"label": "JT", "pattern": str(tok.lower())} for tok in job_titles]
ruler = EntityRuler(nlp, config={
    "phrase_matcher_attr": "LOWER"
})
ruler.add_patterns(patterns)
nlp_updated_implementation.add_pipe(ruler)
```
**Pattern Example**
```
{'label': 'JT', 'pattern': 'chief executives'},
{'label': 'JT', 'pattern': 'chief sustainability officers'},
```

This is exponentially faster than the previous approach that I used. Below is the comparison
```
#This is the initial approach 
%%timeit
nlp_initial_implementation.pipe(df.iloc[:10]['txt'], n_process = len(df.iloc[:10]['txt'])-1, batch_size = 1)
4.86 s ± 93.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

```
#This is the updated approach 
%%timeit
nlp_updated_implementation.pipe(df.iloc[:10]['txt'],  n_process = len(df.iloc[:10]['txt'])-1, batch_size = 1)
900 ms ± 9.86 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit #using nlp with list comprehension is even faster
[nlp_updated_implementation(cont) for cont in df.iloc[:10]['txt']]
169 ms ± 1.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**But using the phrase matcher makes matching ents case sensitive.** 
```
str_x = "Sundar Pichai is the Chief Executive Officer of Google"

print(nlp_initial_implementation(str_x).ents)
(Chief Executive Officer,)

print(nlp_updated_implementation(str_x).ents)
() #returns no matched

```

Is there a way to get case insensitive matches using entity ruler with phrase matcher enabled?

@polm @adrianeboyd

**Reply by bhargav0795** [reply]
Adding to my previous comment, I've built the entity ruler using the following pattern
`{'label': 'JT', 'pattern': 'Chief Executive Officer'},`

**Entity ruler definition**
```
nlp_updated_implementation= spacy.blank("en")
patterns = [{"label": "JT", "pattern": str(tok)} for tok in job_titles]
ruler = EntityRuler(nlp, config={
    "phrase_matcher_attr": "LOWER"
})
ruler.add_patterns(patterns)
nlp_updated_implementation.add_pipe(ruler)
```
**Examples**
```
str_x = "Sundar Pichcai is the CHIEF EXECUTIVE OFFICER of Google"
print(nlp_updated_implementation(str_x).ents)
() #returns no matched but should identify "CHIEF EXECUTIVE OFFICER"

str_y = "Sundar Pichcai is the Chief Executive Officer of Google"
print(nlp_updated_implementation(str_y).ents)
(Chief Executive Officer,)

str_z = "Sundar Pichcai is the chief executive officer of Google"
print(nlp_updated_implementation(str_z).ents)
() #returns no matches but should identify "chief executive officer"
```
In #7585 @polm mentioned that "You can tell the EntityRuler to match on the lower case form of a word so that changes in capitalization don't matter" as long as the `phrase_matcher_attr="LOWER"` flag is enabled. 

Am I doing anything wrong?

**Reply by adrianeboyd** [reply]
For v2 the initialization/configuration looks different:

```python
ruler = EntityRuler(nlp, phrase_matcher_attr="LOWER")
```

Be sure to refer to the v2 docs (https://v2.spacy.io) for v2.x.

**Reply by bhargav0795** [reply]
oh, shoot! It works now. Thanks @adrianeboyd
